Users cannot delete/deactivate themselves or last admin

maxkorp opened the pull request:
https://github.com/rancher/rancher/issues/14818

There are tests for deleting yourself, but not for deleting the last admin, as the only way you could get to that point (past deleting oneself) is a race condition, where adminA deletes adminB at the same time as adminB deletes adminA (or any cycle of the sort).

cjellick commented:
Feedback given in person: simplify down to "cant delete yourself" and add a utility to re-add the admin user from the command line, similar to how we can reset password

maxkorp commented:
Simplified the logic, added `ensure-default-admin` command which will add a default admin if none exists. If one does, it will ensure they still have admin privileges and that their user is active.

cjellick commented on tests/core/test_users.py:
```diff
@@ -0,0 +1,25 @@
+    except ApiError:
+        deleted = False
+
+    assert deleted is False
```
instead of this try, expect, assert paradigm, can you use pytest's
```
    with pytest.raises(ApiError):
```
functionality?

maxkorp replied:
Yup. I also changed the error code returned to a 422, as it was 500ing.

cjellick commented on pkg/api/customization/authn/user_store.go:
```diff
@@ -188,3 +188,43 @@ func (s *userStore) create(apiContext *types.APIContext, schema *types.Schema, d
+func getUser(apiContext *types.APIContext) (string, error) {
+	user := apiContext.Request.Header.Get("Impersonate-User")
+	if user == "" {
+		return "", httperror.NewAPIError(httperror.NotFound, "missing user")
```
If this header is missing at this point, rancher has royally screwed up. Can you change the error code to 500 and a message/code indicating that there was an error authorizing the request.
Its not a 401 or 403 because this is a bug if we get to this point in the code and that header is not present

cjellick commented:
@maxkorp open up a documentation issue in rancher/docs and assign yourself and set it to the 2.1 milestone.
Documentation for this should wind up in the technical faq, close to where we document how to reset the admin password
https://rancher.com/docs/rancher/v2.x/en/faq/technical/#how-can-i-reset-the-admin-password

cjellick commented on app/ensure_admin_user.go:
```diff
@@ -0,0 +1,143 @@
+			for _, u := range admins.Items {
+				users = append(users, u.Name)
+			}
+			return errors.Errorf("%v users were found with %v label. They are %v. Can only reset the default admin password when there is exactly one user with this label.",
```
Bad copy paste of error message from the reset password logic?

maxkorp replied:
Hrm sort of yeah. In this case, there are multiple default admins found (who may or may not be enabled or remain actual admins. 

Do we just want to do it for all of them in that case?  

Obviously there _shouldn't_ be more than one default admin, but we're accommodating for that case in reset-password, so I'm guessing it's come up at some point. I guess technically it would just be multiple users who got _labeled_ in a way indicating them as the default admin.

cjellick commented on app/ensure_admin_user.go:
```diff
@@ -0,0 +1,143 @@
+		DisplayName:        "Default Admin",
+		Username:           "admin",
+		Password:           string(hashedPass),
+		MustChangePassword: true,
```
since the password we are generating is complex and strong and random, no need to require them to change it again. Can set this to false

maxkorp commented on app/ensure_admin_user.go:
```diff
@@ -0,0 +1,143 @@
+		return err
+	}
+
+	admin, err := client.Users("").Create(&v3.User{
```
Aha this will address my above question about labeling as well.

maxkorp replied:
Do we want to address the "multiple users with the name `admin`" while we're here?

maxkorp commented on app/ensure_admin_user.go:
```diff
@@ -0,0 +1,143 @@
+
+func ensureAdminIsEnabled(client v3.Interface, admin v3.User) error {
+	_true := true
+	admin.Enabled = &_true
```
I tried the following, but couldn't get anything to work
```
admin.Enabled = true
// cannot use true (type bool) as type *bool in assignment

admin.Enabled = &true
cannot take the address of true
```

When I did some googling, I found other people with the same issue and generally, the answer seemed to always be some version of the assignment to a temporary var and then dereferencing it.

cjellick commented on tests/core/test_users.py:
```diff
@@ -0,0 +1,25 @@
+    except ApiError:
+        deleted = False
+
+    assert deleted is False
```
@maxkorp why did you mark this resolved? you're still not using pytest.raises

cjellick commented:
LGTM
